core/domain/opportunity_services.py

```diff
@@ -407,6 +407,199 @@ def update_opportunities_with_new_topic_name(topic_id, topic_name):
         exploration_opportunity_summary_list)
 
 
+def get_skill_opportunity_from_model(model):
+    """Returns a SkillOpportunity domain object from a SkillOpportunityModel.
+
+    Args:
+        model: SkillOpportunityModel. The skill opportunity model.
+
+    Returns:
+        SkillOpportunity. The corresponding SkillOpportunity object.
+    """
+
+    return opportunity_domain.SkillOpportunity(
+        model.id, model.skill_description, model.question_count)
+
+
+def get_skill_opportunities(cursor):
+    """Returns a list of skill opportunities available for questions.
+
+    Args:
+        cursor: str or None. If provided, the list of returned entities
+            starts from this datastore cursor. Otherwise, the returned
+            entities start from the beginning of the full list of entities.
+
+    Returns:
+        3-tuple(opportunities, cursor, more). where:
+            opportunities: list(dict). A list of dict of opportunity details.
+            cursor: str or None. A query cursor pointing to the next
+                batch of results. If there are no more results, this might
+                be None.
+            more: bool. If True, there are (probably) more results after
+                this batch. If False, there are no further results after
+                this batch.
+    """
+    page_size = feconf.OPPORTUNITIES_PAGE_SIZE
+    skill_opportunity_models, cursor, more = (
+        opportunity_models.SkillOpportunityModel
+        .get_skill_opportunities(page_size, cursor))
+    opportunities = []
+    for skill_opportunity_model in skill_opportunity_models:
+        skill_opportunity = (
+            get_skill_opportunity_from_model(
+                skill_opportunity_model))
+        opportunities.append(skill_opportunity.to_dict())
+    return opportunities, cursor, more
+
+
+def create_skill_opportunity(skill_id, skill_description):
+    """Creates a skill opportunity domain object as a SkillOpportunityModel
+    entity in the datastore.
+
+    Args:
+        skill_id: str. The skill_id of the opportunity.
+        skill_description: str. The skill_description of the opportunity.
+
+    Raises:
+        Exception: If a SkillOpportunityModel corresponding to the supplied
+            skill_id already exists.
+    """
+    skill_opportunity_model = (
+        opportunity_models.SkillOpportunityModel.get_by_id(skill_id))
+    if skill_opportunity_model:
+        raise Exception(
+            'SkillOpportunity corresponding to skill ID %s already exists.' % (
+                skill_id))
+
+    skill_opportunity = opportunity_domain.SkillOpportunity(
+        skill_id=skill_id,
+        skill_description=skill_description,
+        question_count=0
```

Comment:
Why 0? The skill might already have questions liked to it right?

Reply:
The intent was that this method would be called when a new Skill is created, and since a Skill itself has no knowledge of linked questions (I believe question skill relationships are tracked by Questions or QuestionSkillLinks), I was assuming that question_count would always be 0 here.

Reply:
It's probably safer not to assume that. See if you can compute it directly from the source information in the datastore. This method may be used in other contexts in the future, e.g. regeneration of skill opportunity models.